google/cloud/storage/blob.py

```diff
@@ -2162,14 +2189,18 @@ def _do_upload(
                   **only** response in the multipart case and it will be the
                   **final** response in the resumable case.
         """
-        if if_metageneration_match is None and num_retries is None:
-            # Uploads are only idempotent (safe to retry) if
-            # if_metageneration_match is set. If it is not set, the default
-            # num_retries should be 0. Note: Because retry logic for uploads is
-            # provided by the google-resumable-media-python package, it doesn't
-            # use the ConditionalRetryStrategy class used in other API calls in
-            # this library to solve this problem.
-            num_retries = 0
+
+        # Handle ConditionalRetryPolicy.
+        if isinstance(retry, ConditionalRetryPolicy):
```

Comment:
Will there be cases where `if_metageneration_match is not None `and `retry=None`? Or is it not a concern to this private method? IIUC, it seems like `if_metageneration_match` will be picked up as query params in the upload url eventually, but at the same time we will have retry as `resumable_media.RetryStrategy(max_retries=0)`? Will the conditional retry still be triggered?

Reply:
Yes, `if_metageneration_match` not being `None` and `retry` being `None` is a valid state here. `query_params` inside this conditional _only_ exists for the purpose of evaluating a `ConditionalRetryPolicy`. In any scenario where there is no `ConditionalRetryPolicy`, this code will not fire, but `if_generation_match` and `if_metageneration_match` will still be consumed properly and added to the request later in the process.